### ISSUE TITLE
[Bug] Request API is inconsistent when dealing with multi-value parameters

### ISSUE THREAD

**Issue by yozlet** [Kong/insomnia#1518]
Say that we have multiple parameters with the same name. `request.parameters` looks like:
```json
[
    { "name": "multi", "value": "a" },
    { "name": "multi", "value": "b" },
    { "name": "multi", "value": "c" }
]
```

There are these problems when accessing this multi-value parameter with the existing `context.request.*Parameter` methods:

* `getParameter("multi")` will return `c`, because it only returns the _last_ value of a multi-value parameter
* `setParameter("multi", "d")` will replace the _first_ value, so...
* another call to `getParameter("multi")` will still return `c`
* and `addParameter("multi", "e")` will silently do nothing at all.

Furthermore, there are no methods that allow one to create a multi-value parameter, or to fetch it by name. (However, `getParameters()` does return the entire parameter collection.)

I have the beginnings of a pull request that fixes some of these problems, but they'd either require creation of new `request` methods or changing the behaviour of existing ones (though only changing the behaviour when dealing with multi-value parameters, and if we consider the current behaviour buggy then maybe that's OK?). @gschier: What are your preferences here? Are you OK with changing existing behaviour, or would you prefer something else?

**Comment by gschier** [comment]
Thanks for digging through this issue @yozlet! This was indeed a design mistake early on.

I would like to avoid changing existing behavior if at all possible. Especially since I know that method is used quite frequently.

I just took a look at Node's `querystring` module docs and they use arrays **only** if there are multiple with the same name. I don't necessarily agree with this design 100% but I think it's a familiar enough example that we can do the same.

So, here's what I propose:

- Change `getParameter` to return a string for single matches and an array for multiple
- Change `setParameter` to remove all instances of key and replace with a single new one
- Leave `addParameter(name: string)` as is, since it's working as expected. The method name might be a bit confusing, but the purpose of this one is to only set the parameter if one by that name does not yet exist. 

You don't have to do this if you don't want, but we should also make the same changes to the header methods, as they can also have duplicate entries.

**Comment by yozlet** [comment]
@gschier Agreed on everything with the possible exception of `addParameter()`, since it feels like silent failure and may be hard to debug. Perhaps it could either raise an exception or log a warning to the console?

I already started on this yesterday, with the same `setParameter()` behaviour you've asked for: #1521. I don't know how much time I'll have to complete it, but I'll push my work as I go so that someone else can pick it up if needed.

**Comment by yozlet** [comment]
Also, how about `setParameter(string, array)` for creating/setting a multi-value parameter?

**Comment by wdawson** [comment]
Sorry folks! This was incorrectly tagged as a feature request along with other bugs. I'm going back through and cleaning up, sorry for the inconvenience.
